File: app/src/main/java/com/threecats/livecam/OverlayView.java
/*
 * Copyright (c) 2020 rumburake@gmail.com
 */

package com.threecats.livecam;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.view.View;

public class OverlayView extends View {

    private Matrix matrix;

    public OverlayView(Context context) {
        super(context);
    }

    public OverlayView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    SparseArray<BoxShape> shapeSparseArray = new SparseArray<>();

    public void addShape(int key, BoxShape shape) {
        shapeSparseArray.append(key, shape);
        invalidate();
    }

    public void delShape(int key) {
        shapeSparseArray.delete(key);
        invalidate();
    }

    public void setPreviewRect(RectF previewRect) {
        matrix = new Matrix();
        matrix.setScale( getWidth() / - previewRect.width() , getHeight() / previewRect.height());
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        for (int i = 0; i < shapeSparseArray.size(); ++i) {
            shapeSparseArray.valueAt(i).draw(canvas, matrix);
        }
    }
}

File: app/src/main/java/com/threecats/livecam/OverlayView.kt
/*
 * Copyright (c) 2020 rumburake@gmail.com
 */
package com.threecats.livecam

import android.content.Context
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.RectF
import android.util.AttributeSet
import android.util.SparseArray
import android.view.View

class OverlayView : View {
    private var transformMatrix = Matrix()

    constructor(context: Context?) : super(context) {}
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {}

    var shapeSparseArray = SparseArray<BoxShape>()
    fun addShape(key: Int, shape: BoxShape) {
        shapeSparseArray.append(key, shape)
        invalidate()
    }

    fun delShape(key: Int) {
        shapeSparseArray.delete(key)
        invalidate()
    }

    fun setPreviewRect(previewRect: RectF) {
        transformMatrix.setScale(width / -previewRect.width(), height / previewRect.height())
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        for (i in 0 until shapeSparseArray.size()) {
            shapeSparseArray.valueAt(i).draw(canvas, transformMatrix)
        }
    }
}